app/src/main/java/io/homeassistant/companion/android/util/TLSWebViewClient.kt

```diff
@@ -0,0 +1,119 @@
+package io.homeassistant.companion.android.util
+
+import android.app.Activity
+import android.content.Context
+import android.content.ContextWrapper
+import android.os.Build
+import android.security.KeyChain
+import android.security.KeyChainAliasCallback
+import android.webkit.ClientCertRequest
+import android.webkit.WebView
+import android.webkit.WebViewClient
+import androidx.annotation.RequiresApi
+import io.homeassistant.companion.android.common.data.keychain.KeyChainRepository
+import kotlinx.coroutines.runBlocking
+import java.security.Principal
+import java.security.PrivateKey
+import java.security.cert.X509Certificate
+import javax.inject.Inject
+
+open class TLSWebViewClient @Inject constructor(private var keyChainRepository: KeyChainRepository) : WebViewClient() {
+
+    var isTLSClientAuthNeeded = false
+        private set
+
+    var hasUserDeniedAccess = false
+        private set
+
+    class MyKeyChainCallbackAlias(var context: Context) : KeyChainAliasCallback {
+        var alias: String? = null
+        var ready: Boolean = false
+
+        override fun alias(alias: String?) {
+            this.alias = alias
+            ready = true
+        }
+    }
+
+    private fun getActivity(context: Context?): Activity? {
+        if (context == null) {
+            return null
+        } else if (context is ContextWrapper) {
+            return if (context is Activity) {
+                context
+            } else {
+                getActivity((context).baseContext)
+            }
+        }
+        return null
+    }
+
+    @RequiresApi(Build.VERSION_CODES.M)
+    override fun onReceivedClientCertRequest(view: WebView, request: ClientCertRequest) {
+        // Let the WebViewActivity know the endpoint requires TLS Client Auth
+        isTLSClientAuthNeeded = true
+
+        // Aim to obtain the private key for the whole lifecycle of the WebViewActivity
+        val activity = getActivity(view.context)
+        if (activity == null) return
+
+        var key: PrivateKey?
+        var chain: Array<X509Certificate>?
+
+        // Get the key and the chain (if the user previously chose)
+        runBlocking {
+            key = keyChainRepository.getPrivateKey()
+            chain = keyChainRepository.getCertificateChain()
+        }
```

Comment:
This feels wrong, can we launch this whole process in a coroutine and call `request.proceed` when we are ready?

Reply:
Not an Android/Kotlin so to be sure, I assume something like this would run non-blocking?
```
runBlocking {
    launch {
         <stuff>
```

Reply:
@magicbuffer no, the snippet you pasted won't run non-blocking! See this little [Kotlin playground snippet](https://play.kotlinlang.org/#eyJ2ZXJzaW9uIjoiMS42LjIxIiwicGxhdGZvcm0iOiJqYXZhIiwiYXJncyI6IiIsIm5vbmVNYXJrZXJzIjp0cnVlLCJ0aGVtZSI6ImlkZWEiLCJjb2RlIjoiaW1wb3J0IGtvdGxpbnguY29yb3V0aW5lcy4qXG5cbmZ1biBtYWluKCkge1xuICAgIHJ1bkJsb2NraW5nIHtcbiAgICBcdGxhdW5jaCB7XG5cdFx0ICAgIHByaW50bG4oXCJpbnNpZGUgbGF1bmNoXCIpXG4gICAgICAgICAgICBkZWxheSgxMDAwTClcblx0XHQgICAgcHJpbnRsbihcImluc2lkZSBsYXVuY2ggYWZ0ZXIgMXNcIilcbiAgICAgICAgfVxuXHQgICAgcHJpbnRsbihcImFmdGVyIGxhdW5jaFwiKVxuXHR9XG4gICAgcHJpbnRsbihcImFmdGVyIHJ1bkJsb2NraW5nXCIpXG59In0=) I wrote as a demonstration.

As a solution, I can imagine something like obtaining/constructing a `CoroutineScope` in `TLSWebViewClient`, and calling `launch` on that.

Reply:
I like this idea, I will try to reuse that `lifecycleScope` before attempting to create a new one :+1:

Reply:
It is working like a charm :clap:  thank you!